File: mini-newtonian/src/momentum.jl
# momentum.jl — 动量、冲量、碰撞
# 参考：MIT 8.012 Ch.6, Goldstein Ch.2

## ============================================================
## 单粒子动量
## ============================================================

"""线动量: p = m*v"""
linear_momentum(m::Float64, v::Vec3) = m * v

"""角动量 (绕原点): L = r × p = m(r × v)"""
function angular_momentum(r::Vec3, v::Vec3, m::Float64)
    return m * cross(r, v)
end

"""角动量（绕任意点 origin）"""
function angular_momentum_about(r::Vec3, v::Vec3, m::Float64, origin::Vec3)
    return m * cross(r - origin, v)
end

## ============================================================
## 系统动量
## ============================================================

"""N 体系统总线动量"""
function total_linear_momentum(masses::Vector{Float64}, velocities::Vector{Vec3})
    p_total = Vec3()
    for i in 1:length(masses)
        p_total = p_total + masses[i] * velocities[i]
    end
    return p_total
end

"""N 体系统总角动量（绕原点）"""
function total_angular_momentum(masses::Vector{Float64}, positions::Vector{Vec3}, velocities::Vector{Vec3})
    L_total = Vec3()
    for i in 1:length(masses)
        L_total = L_total + angular_momentum(positions[i], velocities[i], masses[i])
    end
    return L_total
end

"""质心位置"""
function center_of_mass(masses::Vector{Float64}, positions::Vector{Vec3})
    M_total = sum(masses)
    cm = Vec3()
    for i in 1:length(masses)
        cm = cm + masses[i] * positions[i]
    end
    return cm / M_total
end

"""质心速度"""
function center_of_mass_velocity(masses::Vector{Float64}, velocities::Vector{Vec3})
    M_total = sum(masses)
    v_cm = Vec3()
    for i in 1:length(masses)
        v_cm = v_cm + masses[i] * velocities[i]
    end
    return v_cm / M_total
end

"""约化质量: μ = m1*m2 / (m1+m2)"""
reduced_mass(m1::Float64, m2::Float64) = m1 * m2 / (m1 + m2)

## ============================================================
## 冲量
## ============================================================

"""冲量 J = F*Δt（恒力）"""
impulse_constant(F::Vec3, dt::Float64) = F * dt

"""冲量 = Δp = m*(v_f - v_i)（动量变化）"""
impulse_from_momentum(m::Float64, v_i::Vec3, v_f::Vec3) = m * (v_f - v_i)

"""平均力 = 冲量 / 时间"""
average_force_from_impulse(J::Vec3, dt::Float64) = J / dt

"""冲量-动量定理验证"""
function check_impulse_momentum(m::Float64, v_before::Vec3, v_after::Vec3, J::Vec3; tol::Float64=1e-10)
    delta_p = m * (v_after - v_before)
    diff = norm(delta_p - J)
    return diff < tol, diff
end

## ============================================================
## 一维碰撞
## ============================================================

"""
一维完全弹性碰撞
返回 (v1_final, v2_final)
"""
function elastic_collision_1d(m1::Float64, v1::Float64, m2::Float64, v2::Float64)
    v1f = (m1 - m2) / (m1 + m2) * v1 + 2 * m2 / (m1 + m2) * v2
    v2f = 2 * m1 / (m1 + m2) * v1 + (m2 - m1) / (m1 + m2) * v2
    return v1f, v2f
end

"""
一维完全非弹性碰撞（粘在一起）
返回共同末速度
"""
function inelastic_collision_1d(m1::Float64, v1::Float64, m2::Float64, v2::Float64)
    v_common = (m1 * v1 + m2 * v2) / (m1 + m2)
    return v_common
end

"""
碰撞恢复系数: e = |v2f - v1f| / |v2i - v1i|
e=1 弹性, e=0 完全非弹性
"""
function coefficient_of_restitution(v1i::Float64, v2i::Float64, v1f::Float64, v2f::Float64)
    return abs(v2f - v1f) / abs(v2i - v1i)
end

"""
给定恢复系数 e 的一维碰撞
"""
function collision_with_restitution(m1::Float64, v1::Float64, m2::Float64, v2::Float64, e::Float64)
    v_cm = (m1 * v1 + m2 * v2) / (m1 + m2)
    # 在质心系中速度乘以 -e
    v1_cm = v1 - v_cm
    v2_cm = v2 - v_cm
    v1f_cm = -e * v1_cm
    v2f_cm = -e * v2_cm
    return v1f_cm + v_cm, v2f_cm + v_cm
end

## ============================================================
## 三维碰撞
## ============================================================

"""
三维完全弹性碰撞（光滑球）
假定碰撞时沿法线方向交换动量，切向不变
normal: 碰撞法线方向（从球1中心指向球2中心）
"""
function elastic_collision_3d(m1::Float64, v1::Vec3, m2::Float64, v2::Vec3, normal::Vec3)
    n = normalize(normal)
    # 分解为法向和切向分量
    v1n = dot(v1, n)
    v1t = v1 - v1n * n
    v2n = dot(v2, n)
    v2t = v2 - v2n * n

    # 法向弹性碰撞
    v1nf, v2nf = elastic_collision_1d(m1, v1n, m2, v2n)

    return v1t + v1nf * n, v2t + v2nf * n
end

## ============================================================
## 火箭方程（变质量系统）
## ============================================================

"""
齐奥尔科夫斯基火箭方程
Δv = ve * ln(m0 / mf)
ve: 排气速度, m0: 初始质量, mf: 最终质量
"""
function tsiolkovsky_delta_v(ve::Float64, m0::Float64, mf::Float64)
    return ve * log(m0 / mf)
end

"""
给定 Δv 需要的质量比
"""
function mass_ratio_for_delta_v(delta_v::Float64, ve::Float64)
    return exp(delta_v / ve)
end

File: mini-newtonian/src/momentum.c
/**
 * momentum.c - Linear/angular momentum, impulse, collisions
 * Each function implements an independent physics knowledge point.
 * Course: MIT 8.012 Ch.6, Goldstein Ch.2-3, Kleppner & Kolenkow Ch.4
 */
#include "momentum.h"
#include <math.h>

/* === L1: Single-particle momentum ===
 * p = m*v
 * L = r x p = m(r x v)
 */
Vec3 linear_momentum(double m, Vec3 v) {
    return vec3_scale(v, m);
}
Vec3 angular_momentum(Vec3 r, Vec3 v, double m) {
    return vec3_scale(vec3_cross(r, v), m);
}
Vec3 angular_momentum_about(Vec3 r, Vec3 v, double m, Vec3 origin) {
    return angular_momentum(vec3_sub(r, origin), v, m);
}

/* === L2: System momentum === */
Vec3 total_linear_momentum(double masses[], Vec3 velocities[], int n) {
    Vec3 p_total = vec3_zero();
    for (int i = 0; i < n; i++)
        p_total = vec3_add(p_total, linear_momentum(masses[i], velocities[i]));
    return p_total;
}
Vec3 total_angular_momentum(double masses[], Vec3 positions[], Vec3 velocities[], int n) {
    Vec3 L_total = vec3_zero();
    for (int i = 0; i < n; i++)
        L_total = vec3_add(L_total, angular_momentum(positions[i], velocities[i], masses[i]));
    return L_total;
}

/* Center of mass: R_cm = sum(m_i*r_i) / sum(m_i)
 * Knowledge point: CM moves as if all mass and external forces concentrated there.
 * Theorem: F_ext = M * a_cm (Center of Mass Theorem)
 */
Vec3 center_of_mass(double masses[], Vec3 positions[], int n) {
    Vec3 cm = vec3_zero();
    double M_total = 0.0;
    for (int i = 0; i < n; i++) {
        cm = vec3_add(cm, vec3_scale(positions[i], masses[i]));
        M_total += masses[i];
    }
    if (M_total < 1e-15) return vec3_zero();
    return vec3_div(cm, M_total);
}
Vec3 center_of_mass_velocity(double masses[], Vec3 velocities[], int n) {
    Vec3 v_cm = vec3_zero();
    double M_total = 0.0;
    for (int i = 0; i < n; i++) {
        v_cm = vec3_add(v_cm, vec3_scale(velocities[i], masses[i]));
        M_total += masses[i];
    }
    if (M_total < 1e-15) return vec3_zero();
    return vec3_div(v_cm, M_total);
}

/* Reduced mass: mu = m1*m2 / (m1+m2)
 * Knowledge point: Two-body problem reduces to one-body with reduced mass.
 * Course: Goldstein Ch.3.1
 */
double reduced_mass(double m1, double m2) {
    return m1 * m2 / (m1 + m2);
}

/* === L4: Impulse J = F*dt = Delta(p) ===
 * Impulse-Momentum Theorem: integral(F dt) = p_final - p_initial
 */
Vec3 impulse_constant_force(Vec3 F, double dt) {
    return vec3_scale(F, dt);
}
Vec3 impulse_from_momentum_change(double m, Vec3 v_initial, Vec3 v_final) {
    return vec3_scale(vec3_sub(v_final, v_initial), m);
}
Vec3 average_force_from_impulse(Vec3 impulse, double dt) {
    if (fabs(dt) < 1e-15) return vec3_zero();
    return vec3_div(impulse, dt);
}
int verify_impulse_momentum(double m, Vec3 v_before, Vec3 v_after,
                            Vec3 impulse, double tol) {
    Vec3 delta_p = vec3_scale(vec3_sub(v_after, v_before), m);
    double diff = vec3_norm(vec3_sub(delta_p, impulse));
    return (diff < tol) ? 1 : 0;
}

/* === L6: 1D Collisions ===
 * Elastic: both momentum and kinetic energy conserved
 *   v1f = ((m1-m2)/(m1+m2))*v1 + (2*m2/(m1+m2))*v2
 *   v2f = (2*m1/(m1+m2))*v1 + ((m2-m1)/(m1+m2))*v2
 */
void elastic_collision_1d(double m1, double v1, double m2, double v2,
                          double *v1f, double *v2f) {
    double M = m1 + m2;
    *v1f = (m1 - m2) / M * v1 + 2.0 * m2 / M * v2;
    *v2f = 2.0 * m1 / M * v1 + (m2 - m1) / M * v2;
}

/* Perfectly inelastic: stick together, v_common = (m1*v1 + m2*v2)/(m1+m2)
 * Kinetic energy is NOT conserved (maximum loss), only momentum conserved.
 */
double inelastic_collision_1d(double m1, double v1, double m2, double v2) {
    return (m1 * v1 + m2 * v2) / (m1 + m2);
}

/* Coefficient of restitution: e = |v2f - v1f| / |v2i - v1i|
 * e = 1: perfectly elastic
 * e = 0: perfectly inelastic
 * 0 < e < 1: partially elastic (real collisions)
 */
double coefficient_of_restitution(double v1i, double v2i, double v1f, double v2f) {
    double rel_i = fabs(v2i - v1i);
    double rel_f = fabs(v2f - v1f);
    if (rel_i < 1e-15) return 1.0;
    return rel_f / rel_i;
}

/* Collision with arbitrary restitution e (0 <= e <= 1)
 * Uses center-of-mass frame: v_cm' = -e * v_cm
 */
void collision_with_restitution(double m1, double v1, double m2, double v2,
                                double e, double *v1f, double *v2f) {
    double v_cm = (m1 * v1 + m2 * v2) / (m1 + m2);
    double v1_cm = v1 - v_cm;
    double v2_cm = v2 - v_cm;
    *v1f = -e * v1_cm + v_cm;
    *v2f = -e * v2_cm + v_cm;
}

/* Energy loss in partially inelastic collision
 * Delta(E) = E_final - E_initial (negative for energy loss)
 */
double collision_energy_loss(double m1, double v1i, double m2, double v2i,
                             double v1f, double v2f) {
    double E_i = 0.5 * m1 * v1i * v1i + 0.5 * m2 * v2i * v2i;
    double E_f = 0.5 * m1 * v1f * v1f + 0.5 * m2 * v2f * v2f;
    return E_f - E_i;
}

/* === L6: 3D Collisions ===
 * Smooth sphere elastic collision in 3D.
 * Normal components exchange elastically, tangential components unchanged.
 * normal: unit vector from sphere1 center to sphere2 center at contact.
 */
void elastic_collision_3d(double m1, Vec3 v1, double m2, Vec3 v2,
                          Vec3 collision_normal,
                          Vec3 *v1f, Vec3 *v2f) {
    Vec3 n = vec3_normalize(collision_normal);
    double v1n = vec3_dot(v1, n);
    double v2n = vec3_dot(v2, n);
    Vec3 v1t = vec3_sub(v1, vec3_scale(n, v1n));
    Vec3 v2t = vec3_sub(v2, vec3_scale(n, v2n));
    double v1nf, v2nf;
    elastic_collision_1d(m1, v1n, m2, v2n, &v1nf, &v2nf);
    *v1f = vec3_add(v1t, vec3_scale(n, v1nf));
    *v2f = vec3_add(v2t, vec3_scale(n, v2nf));
}

/* === L7: Rocket Equation (Tsiolkovsky 1903) ===
 * Delta_v = v_e * ln(m0/mf)
 * Knowledge points:
 *   1. Variable mass system: thrust from expelled propellant
 *   2. Logarithmic dependence: diminishing returns from more fuel
 *   3. Mass ratio = exp(Delta_v / v_e): exponential mass penalty
 * Course: MIT 8.012 Ch.6.4, Kleppner & Kolenkow Ch.4.6
 */
double tsiolkovsky_delta_v(double ve, double m0, double mf) {
    if (mf <= 0.0 || m0 <= mf) return 0.0;
    return ve * log(m0 / mf);
}
double mass_ratio_for_delta_v(double delta_v, double ve) {
    return exp(delta_v / ve);
}
double rocket_mass_flow_rate(double thrust, double ve) {
    if (fabs(ve) < 1e-15) return 0.0;
    return thrust / ve;
}
double rocket_final_mass(double m0, double burn_time, double mass_rate) {
    double mf = m0 - mass_rate * burn_time;
    return (mf > 0.0) ? mf : 0.0;
}

/* === L4: Conservation law verification === */
int check_momentum_conservation(Vec3 p_initial, Vec3 p_final, double tol) {
    return vec3_norm(vec3_sub(p_final, p_initial)) < tol ? 1 : 0;
}
int check_angular_momentum_conservation(Vec3 L_initial, Vec3 L_final, double tol) {
    return vec3_norm(vec3_sub(L_final, L_initial)) < tol ? 1 : 0;
}
int check_energy_conservation(double E_initial, double E_final, double tol) {
    return fabs(E_final - E_initial) < tol * fmax(fabs(E_initial), 1.0) ? 1 : 0;
}
